Plutarch/Rational.hs

```diff
@@ -38,106 +41,128 @@ import Plutarch.Builtin (
   pforgetData,
   pfromDataImpl,
  )
-import Plutarch.Integer (PInteger, PIntegral (pdiv, pmod))
-import Plutarch.List (PListLike (pcons, phead, pnil, ptail), pmap)
-import Plutarch.Pair (PPair (..))
-import Plutarch.Show (PShow (pshow'), pshow)
-import Plutarch.Trace (ptraceError)
-import Plutarch.Unsafe (punsafeCoerce)
+import Plutarch.Integer (PInteger, pdiv, pmod)
+import Plutarch.Lift (pconstant)
+import Plutarch.List (pcons, phead, pnil, ptail)
+import Plutarch.NonZero (PNonZero, ptryNonZero)
+import Plutarch.Pair (PPair (PPair))
+import Plutarch.Show (PShow, pshow, pshow')
+import Plutarch.TermCont (tcont, unTermCont)
+import Plutarch.Unsafe (punsafeCoerce, punsafeDowncast)
 
 data PRational s
-  = PRational (Term s PInteger) (Term s PInteger)
+  = PRational (Term s PInteger) (Term s PNonZero)
   deriving stock (GHC.Generic)
-  deriving anyclass (Generic, PlutusType)
+  deriving anyclass (Generic, PEq)
+
+{- | NOTE: This instance may delegate the nonzero invariant check during construction.
+
+As such, the constructor is not exported and users are expected to use 'prational', 'prationalLazy', and
+'punrational' instead.
+-}
+instance PlutusType PRational where
+  type PInner PRational b = (PInteger :--> PNonZero :--> b) :--> b
+  pcon' (PRational x y) = plam $ \f -> f # x # y
+  pmatch' ratF f = ratF #$ plam $ \a b -> f $ PRational a b
+
+{- | Builds a rational with lazy non-zero denominator verification.
+
+i.e, if the denominator was zero, it'll error only after 'pmatch'ing on the resulting rational.
+-}
+prationalLazy :: Term s PInteger -> Term s PNonZero -> Term s PRational
```

Comment:
This isn't lazy anymore

Reply:
removed, will figure out optimizations later as you said